password/src/Data/Password/Validate.hs

```diff
@@ -126,122 +143,139 @@ data CharacterCategory
   -- ^ ASCII digits
   deriving (Eq, Ord, Show)
 
--- | Convert 'CharacterCategory' into associated predicate function
+-- | Convert a 'CharacterCategory' into its associated predicate function
+--
+-- @since 2.1.0.0
 categoryToPredicate :: CharacterCategory -> (Char -> Bool)
 categoryToPredicate = \case
   Uppercase -> isAsciiUpper
   Lowercase -> isAsciiLower
   Special -> isSpecial
   Digit -> isDigit
 
--- | Possible reason of 'Password' being invalid
+-- | Possible reason for a 'Password' to be invalid.
+--
+-- @since 2.1.0.0
 data InvalidReason
   = PasswordTooShort !Int !Int
   -- ^ Length of 'Password' is too short.
   --
-  -- Expected at least 'Int' characters but the actual length is 'Int'
+  -- @PasswordTooShort expected found@
   | PasswordTooLong !Int !Int
   -- ^ Length of 'Password' is too long.
   --
-  -- Expected at maximum of 'Int' characters, but the actual length is 'Int'
-  | NotEnoughReqChars CharacterCategory !Int !Int
+  -- @PasswordTooLong expected found@
+  | NotEnoughReqChars !CharacterCategory !Int !Int
   -- ^ 'Password' does not contain required number of characters.
   --
-  -- Expected at least 'Int' characters of 'CharacterCategory' but the password only
-  -- contains 'Int'
+  -- @NotEnoughReqChars category expected found@
   | InvalidCharacters !Text
   -- ^ 'Password' contains characters that cannot be used
-  | InvalidPasswordPolicy [InvalidPolicyReason]
-  -- ^ 'PasswordPolicy' is invalid
   deriving (Eq, Ord, Show)
 
 -- | Possible reason of 'PasswordPolicy' being invalid
+--
+-- @since 2.1.0.0
 data InvalidPolicyReason
   = InvalidLength !Int !Int
   -- ^ Value of 'minimumLength' is bigger than 'maximumLength'
+  --
+  -- @InvalidLength min max@
   | MaxLengthBelowZero !Int
-  -- ^ Value of 'maximumLength' is less than zero
+  -- ^ Value of 'maximumLength' is zero or less
   | InvalidCharSetPredicate !CharacterCategory !Int
   -- ^ 'CharSetPredicate' does not return 'True' for a 'CharacterCategory' that
   -- requires at least 'Int' characters in the password
   deriving (Eq, Ord, Show)
 
+-- | Result of validating a 'Password'.
+--
+-- Note that if the 'PasswordPolicy' is invalid, this will never return 'ValidPassword'.
+--
+-- @since 2.1.0.0
+data ValidationResult
+  = ValidPassword
+  -- ^ The 'Password' conforms to the validation parameters
+  | InvalidPassword [InvalidReason]
+  -- ^ 'Password' failed to be validated for the given reasons
+  | InvalidPolicy [InvalidPolicyReason]
+  -- ^ 'PasswordPolicy' is invalid
```

Comment:
Instead of having this error case, we could also just have a smart constructor for `PasswordPolicy`, but I don't really feel strongly either way.  It would be somewhat annoying to go back and update all the code for this.  I'm completely fine without not doing this.

Reply:
This is a good point for usability, though! I'll see what I can do.

I'm also thinking about maybe adding a QuasiQuote function to have the `PasswordPolicy` checked at compile time, so you don't have to handle a failure case.

Reply:
Hmmm, on second thought, a smart constructor wouldn't be very readable and/or nice to use, I think.

My reasoning:
- It would still require the user to handle the failure case, which the user would also do with `validatePassword`
- If the user would like to make sure the policy is valid, they can use `validatePasswordPolicy`
- A function that takes 5 `Int`s is very prone to mistakes

Reply:
On third thought, I think I've found a pretty user-friendly way of doing it. I've made preparations for it and will hopefully have time next weekend to finish it up.

Reply:
Ok, added `ValidPasswordPolicy`.